Add `make linkcheck` step, to check external links [#29]

genehack opened the pull request:
## Description of proposed changes

Adds a `make linkcheck` step to the Sphinx CI.

Note that this blocks on the various PRs attached to nextstrain/docs.nextstrain.org#14 — those PRs should be merged before this one.

Update to add: I realized this workflow is used by more than `docs.nextstrain.org`, so before this can be merged, the following repos need to be audited for broken links:

- [x] Audit augur -- nextstrain/augur#1570
- [x] Audit auspice --  nextstrain/auspice#1808
- [x] Audit cli -- nextstrain/cli#389
- [x] Audit ncov -- nextstrain/ncov#1137


<!-- What is the goal of this pull request? What does this pull request change? -->

## Related issue(s)

Closes #29 
nextstrain/docs.nextstrain.org#14

<!--
Link any related issues here. Use GitHub's special keywords if appropriate¹.
Type `#` followed the name of an issue and GitHub will auto-suggest the issue number for you.

¹ https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests
-->

## Checklist

<!--
Make sure checks are successful at the bottom of the PR.

If applicable, add:
- any changes to existing tests
- any additional manual testing to confirm changes

Please add a note if you need help with adding tests.
-->

- [x] Checks pass

<!-- 🙌 Thank you for contributing to Nextstrain! ✨ -->

joverlee521 commented on .github/workflows/docs-ci.yaml:
```diff
@@ -82,3 +82,6 @@ jobs:
       - run: make ${{ inputs.make-target }}
         working-directory: ${{ inputs.docs-directory }}
+
+      - run: SPHINXOPTS="-W --keep-going" make linkcheck ${{ inputs.make-target }}
```
_non-blocking question_ 

Is `SPHINXOPTS="-W --keep-going"` meant to override the default envvar set in the workflow? 
https://github.com/nextstrain/.github/blob/429f042333c3cbfd2587d2942d283a57d3d252cc/.github/workflows/docs-ci.yaml#L48

genehack replied:
> _non-blocking question_
> 
> Is `SPHINXOPTS="-W --keep-going"` meant to override the default envvar set in the workflow?
> 
> https://github.com/nextstrain/.github/blob/429f042333c3cbfd2587d2942d283a57d3d252cc/.github/workflows/docs-ci.yaml#L48

TBPH, I copied that invocation [from Tom's original note in the issue](https://github.com/nextstrain/.github/issues/29#issue-1442955952), and I didn't notice there was a default setting. Experimenting locally, `-n` doesn't seem to make a difference one way or the other, so I'll remove the env var in this step.

joverlee521 reviewed (changes requested):
I think we'd want to add the new `make linkcheck` step the the `build-conda` job as well?

genehack commented:
> I think we'd want to add the new `make linkcheck` step the the `build-conda` job as well?

Good call, done.

Also removed env-var and adjusted use of `make-target`.

genehack commented:
Updated description above to note that before this can land, everything that uses it (listed above) needs to be manually audited to get `make linkcheck` to run cleanly. Links to open issues above.

victorlin commented on .github/workflows/docs-ci.yaml:
```diff
@@ -82,3 +85,6 @@ jobs:
       - run: make ${{ inputs.make-target }}
         working-directory: ${{ inputs.docs-directory }}
+
+      - run: make linkcheck
+        working-directory: ${{ inputs.docs-directory }}
```
> I realized this workflow is used by more than `docs.nextstrain.org`, so before this can be merged, the following repos need to be audited for broken links:

An alternative is to make this opt-in. This can be achieved by specifying a new workflow input (e.g. `linkcheck-target` and conditioning this step on its presence:

```
- if: inputs.linkcheck-target
  run: make ${{ linkcheck-target }}
  working-directory: ${{ inputs.docs-directory }}
```

Usage would look like:

```
  docs-ci:
    uses: nextstrain/.github/.github/workflows/docs-ci.yaml@master
    with:
      environment-file: conda.yml
      docs-directory: .
      make-target: html
      linkcheck-target: linkcheck
```

genehack replied:
I've got momentum on this now, I'll just keep rolling through the other two repos...

genehack commented:
This is now ready to merge — all the affected downstream repos should have clean `make linkcheck` runs in their doc trees — but I'm going to wait until Monday to actually push the button, just in case something goes pear-shaped after all.

genehack commented:
> This is now ready to merge — all the affected downstream repos should have clean `make linkcheck` runs in their doc trees — but I'm going to wait until Monday to actually push the button, just in case something goes pear-shaped after all.

I also validated a branch in each of the five repos that consume this PI action, changing `docs-ci.yaml` to point to this branch, and verifying that the CI remained green.
